server/src/server.rs

```diff
@@ -42,63 +42,42 @@ impl CubebDeviceCollectionManager {
     }
 
     fn register(
-        &mut self,
+        &self,
         context: &cubeb::Context,
-        server: &Rc<RefCell<DeviceCollectionChangeCallback>>,
+        server: &Rc<DeviceCollectionChangeCallback>,
         devtype: cubeb::DeviceType,
     ) -> cubeb::Result<()> {
-        server.borrow_mut().devtype.insert(devtype);
         let mut servers = self.servers.lock().unwrap();
-        let do_register = servers.is_empty();
-        if !servers.iter().any(|s| Rc::ptr_eq(s, server)) {
-            servers.push(server.clone());
-        }
-        if do_register {
+        if servers.is_empty() {
             self.internal_register(context, true)?;
         }
+        servers.push((server.clone(), devtype));
         Ok(())
     }
 
     fn unregister(
-        &mut self,
+        &self,
         context: &cubeb::Context,
-        server: &Rc<RefCell<DeviceCollectionChangeCallback>>,
+        server: &Rc<DeviceCollectionChangeCallback>,
         devtype: cubeb::DeviceType,
     ) -> cubeb::Result<()> {
-        let do_remove = {
-            server.borrow_mut().devtype.remove(devtype);
-            server.borrow().devtype.is_empty()
-        };
         let mut servers = self.servers.lock().unwrap();
-        if do_remove {
-            servers.retain(|s| !Rc::ptr_eq(s, server));
-        }
+        servers.retain(|(s, d)| !Rc::ptr_eq(s, server) || d != &devtype);
         if servers.is_empty() {
             self.internal_register(context, false)?;
         }
         Ok(())
     }
 
     fn unregister_server(
```

Comment:
Somehow my comment here is lost.

I think we can entirely remove this function now. We can use `self.unregister(context, server, cubeb::DeviceType::all())` directly in `CubebServer::drop`

Reply:
Good idea, I'll push this in a follow up.

Reply:
https://github.com/mozilla/audioipc/commit/44ade28aa69e90103318f2ad501a6962c9413985